lib/ClangImporter/SwiftDeclSynthesizer.cpp

```diff
@@ -1290,10 +1335,95 @@ synthesizeEnumRawValueConstructorBody(AbstractFunctionDecl *afd,
                                      /*implicit*/ true);
   assign->setType(TupleType::getEmpty(ctx));
 
+  // Check if the enum is frozen (closed). If so, we need to validate
+  // that the raw value is one of the declared cases.
+  bool isFrozen = enumDecl->getAttrs().hasAttribute<FrozenAttr>();
+
+  if (isFrozen) {
+    // For frozen enums, generate a switch statement to validate the raw value
+    SmallVector<CaseStmt *, 8> cases;
+
+    // Build cases for each enum element
+    for (auto *elt : enumDecl->getAllElements()) {
+      // Get the raw value literal for this element
+      auto rawValueExpr = elt->getStructuralRawValueExpr();
+      if (!rawValueExpr)
+        continue;
+
+      // Clone the raw value expression for pattern matching
+      auto *litExpr = cloneRawLiteralExpr(ctx, cast<LiteralExpr>(rawValueExpr));
+      auto *litPat = ExprPattern::createImplicit(ctx, litExpr, ctorDecl);
+
+      // Create case label
+      auto labelItem = CaseLabelItem(litPat);
+
+      // Create the assignment statement for this case
+      auto *caseBodySelfRef = new (ctx) DeclRefExpr(selfDecl, DeclNameLoc(),
+                                                     /*implicit*/ true);
+      caseBodySelfRef->setType(LValueType::get(selfDecl->getTypeInContext()));
+
+      auto *caseParamRef = new (ctx) DeclRefExpr(param, DeclNameLoc(),
+                                                  /*implicit*/ true);
+      caseParamRef->setType(param->getTypeInContext());
+
+      auto *caseArgList = ArgumentList::forImplicitUnlabeled(ctx, {caseParamRef});
+      auto *caseReinterpretCastRef =
+          new (ctx) DeclRefExpr(concreteDeclRef, DeclNameLoc(), /*implicit*/ true);
+      caseReinterpretCastRef->setType(
+          FunctionType::get({FunctionType::Param(rawTy)}, enumTy, info));
+
+      auto caseReinterpreted =
+          CallExpr::createImplicit(ctx, caseReinterpretCastRef, caseArgList);
+      caseReinterpreted->setType(enumTy);
+      caseReinterpreted->setThrows(nullptr);
+
+      auto *caseAssign = new (ctx) AssignExpr(caseBodySelfRef, SourceLoc(),
+                                               caseReinterpreted, /*implicit*/ true);
+      caseAssign->setType(TupleType::getEmpty(ctx));
+
+      auto *caseBody = BraceStmt::create(ctx, SourceLoc(), {caseAssign},
+                                         SourceLoc(), /*implicit*/ true);
+
+      auto *caseStmt = CaseStmt::createImplicit(ctx, CaseParentKind::Switch,
```

Comment:
Great contribution! You could save a few bytes by generating one case statement with each known case as a series of labels, as their bodies are the same (reinterpret cast)

equivalent to the source:
```
switch rawValue {
case .first, .second, third:
  reinterpretCast
default:
  return nil
}
```

function merging may consolidate these blocks on the back end but better not to emit them in the first place